rllib/utils/replay_buffers/episode_replay_buffer.py

```diff
@@ -0,0 +1,508 @@
+            if T == batch_length_T:
+                # We may have overfilled this row: Clip trajectory at the end.
+                observations[B] = observations[B][:batch_length_T]
+                actions[B] = actions[B][:batch_length_T]
+                rewards[B] = rewards[B][:batch_length_T]
+                # Start filling the next row.
+                B += 1
+                T = 0
+
+            # Use next sampled episode/ts pair to fill the row.
+            idx_cursor += 1
+
+        # Update our sampled counter.
+        self.sampled_timesteps += batch_size_B * batch_length_T
+
+        # TODO: Return SampleBatch instead of this simpler dict.
+        ret = {
+            "obs": np.array(observations),
+            "actions": np.array(actions),
+            "rewards": np.array(rewards),
+            "is_first": np.array(is_first),
+            "is_last": np.array(is_last),
+            "is_terminated": np.array(is_terminated),
+            "is_truncated": np.array(is_truncated),
+        }
+
+        return ret
+
+    def get_num_episodes(self) -> int:
+        """Returns the number of episodes (completed or truncated) stored in the buffer.
+        """
+        return len(self.episodes)
+
+    def get_num_timesteps(self) -> int:
+        """Returns the number of individual timesteps stored in the buffer."""
+        return len(self._indices)
+
+    def get_sampled_timesteps(self) -> int:
+        """Returns the number of timesteps that have been sampled in buffer's lifetime.
+        """
+        return self.sampled_timesteps
+
+    @override(ReplayBufferInterface)
+    def get_state(self) -> Dict[str, Any]:
+        return {
+            "episodes": [eps.get_state() for eps in self.episodes],
+            "episode_id_to_index": list(self.episode_id_to_index.items()),
+            "_num_episodes_ejected": self._num_episodes_ejected,
+            "_indices": self._indices,
+            "size": self.size,
+            "sampled_timesteps": self.sampled_timesteps,
+        }
+
+    @override(ReplayBufferInterface)
+    def set_state(self, state) -> None:
+        assert state[0][0] == "episodes"
+        self.episodes = deque([
+            _Episode.from_state(eps_data) for eps_data in state[0][1]
+        ])
+        assert state[1][0] == "episode_id_to_index"
+        self.episode_id_to_index = dict(state[1][1])
+        assert state[2][0] == "_num_episodes_ejected"
+        self._num_episodes_ejected = state[2][1]
+        assert state[3][0] == "_indices"
+        self._indices = state[3][1]
+        assert state[4][0] == "size"
+        self.size = state[4][1]
+        assert state[5][0] == "sampled_timesteps"
+        self.sampled_timesteps = state[5][1]
+
+
+# TODO (sven): Make this EpisodeV3 - replacing EpisodeV2 - to reduce all the
+#  information leakage we currently have in EpisodeV2 (policy_map, worker, etc.. are
+#  all currently held by EpisodeV2 for no good reason).
+class _Episode:
```

Comment:
I'm skipping on _Episode for now because I don't assume that we'll have to maintain it.
This assumes that you'll go ahead and change EpisodeV2 to gain EpisodeV3 so that we can review such a change in an upcoming PR.

Reply:
Exactly! Let's consider `_Episode` to be a temporary solution, from with we will derive a new and improved "EpisodeV3" (as well as from `EpisodeV2`).